src/unix/os390-syscalls.c

```diff
@@ -0,0 +1,291 @@
+            const struct dirent **)) {
+  struct dirent **nl;
+  struct dirent *dirent;
+  unsigned count;
+  size_t allocated;
+  DIR *mdir;
+
+  nl = NULL;
+  count = 0;
+  allocated = 0;
+  mdir = opendir(maindir);
+  if (!mdir)
+    return -1;
+
+  while (1) {
+    dirent = readdir(mdir);
+    if (!dirent)
+      break;
+    if (!filter || filter(dirent)) {
+      struct dirent *copy;
+      copy = uv__malloc(sizeof(*copy));
+      if (!copy) {
+        while (count) {
+          dirent = nl[--count];
+          uv__free(dirent);
+        }
+        uv__free(nl);
+        closedir(mdir);
+        errno = ENOMEM;
+        return -1;
+      }
+      memcpy(copy, dirent, sizeof(*copy));
+
+      nl = uv__realloc(nl, count+1);
+      nl[count++] = copy;
+    }
+  }
+
+  qsort(nl, count, sizeof(struct dirent *),
+       (int (*)(const void *, const void *))compar);
+
+  closedir(mdir);
+
+  *namelist = nl;
+  return count;
+}
+
+static int isfdequal(const struct pollfd* first,
+                     const struct pollfd* second) {
+  return first->fd == second->fd ? 0 : 1;
+}
+
+static struct pollfd* findpfd(int epfd, int fd, int events) {
+  struct epoll_list *lst;
+  struct pollfd pfd;
+  struct pollfd* found;
+
+  pfd.fd = fd;
+  pfd.events = events;
+  lst = global_epoll_list[epfd];
+  if (events == 0)
+    return lfind(&pfd, &lst->items[0], &lst->size,
+                  sizeof(pfd), isfdequal);
+  else
+    return lsearch(&pfd, &lst->items[0], &lst->size,
+                  sizeof(pfd), isfdequal);
+}
+
+static void epoll_init() {
+  uv_mutex_init(&global_epoll_lock);
+}
+  
+int epoll_create1(int flags) {
+  struct epoll_list* list;
+  int index;
+
+  uv_once(&once, epoll_init);
+  uv_mutex_lock(&global_epoll_lock);
+  list = uv__malloc(sizeof(*list));
+  index = number_of_epolls++;
+  global_epoll_list[index] = list;
+  uv_mutex_unlock(&global_epoll_lock);
+
+  /* initialize list */
+  memset(list, 0, sizeof(*list));
+  if (uv_mutex_init(&list->lock)) {
+    errno = ENOLCK;
+    return -1;
+  }
+
+  list->size = 0;
+  return index; 
+}
+
+int epoll_ctl(int epfd, int op, int fd, struct epoll_event *event) {
+  struct epoll_list *lst;
+
+  uv_mutex_lock(&global_epoll_lock);
+  lst = global_epoll_list[epfd];
+  uv_mutex_lock(&lst->lock);
+  uv_mutex_unlock(&global_epoll_lock);
+
+  if(op == EPOLL_CTL_DEL) {
+    struct pollfd* found;
+
+    found = findpfd(epfd, fd, 0); 
+    if (found != NULL)
+      memcpy(found,
+             found+1,
+             (&lst->items[lst->size--] - found - 1) * sizeof(*found));
+    uv_mutex_unlock(&lst->lock);
+
+    if (found == NULL)
+      return ENOENT;
+
+  }
+  else if(op == EPOLL_CTL_ADD) {
+    struct pollfd* found;
+    size_t before;
+    size_t after;
+
+    if (lst->size == MAX_ITEMS_PER_EPOLL - 1) {
+      uv_mutex_unlock(&lst->lock);
+      return ENOMEM;
+    }
+
+    before = lst->size; 
+    found = findpfd(epfd, fd, event->events);
+    after = lst->size; 
```

Comment:
why is this needed? can `findpfd` modify the list? if so it needs a better name, finding something shouldn't alter the size of the list.